merlin/models/tf/models/base.py

```diff
@@ -480,10 +485,73 @@ def test_step(self, data):
             outputs = self.pre_eval_topk.call_outputs(outputs)
 
         self.compute_loss(x, outputs.targets, outputs.predictions, sample_weight)
-        metrics = self.compute_metrics(x, outputs.targets, outputs.predictions, sample_weight)
+        metrics = self.compute_metrics(
+            x, outputs.targets, outputs.predictions, sample_weight, training=False
+        )
+        # Adding regularization loss to metrics
+        metrics["regularization_loss"] = tf.reduce_sum(cast_losses_to_common_dtype(self.losses))
 
         return metrics
 
+    @tf.function
+    def compute_metrics(
+        self,
+        x: tf.Tensor,
+        y: tf.Tensor,
+        y_pred: tf.Tensor,
+        sample_weight: tf.Tensor,
+        training: bool,
+    ) -> Dict[str, tf.Tensor]:
+        """Overrides default Keras Model.compute_metrics()
+           to compute metrics each N steps (and not for all steps)
+           during training
+
+        Parameters
+        ----------
+        x : tf.Tensor
+            Input features. Not used, and kept only to align with the signature
+            of super Model.compute_metrics()
+        y : tf.Tensor
+            Target values
+        y_pred : tf.Tensor
+            Predictions values
+        sample_weight : tf.Tensor
+            Sample weights for weighted metrics
+        training : bool
+            Flag that indicates if metrics are being computed during
+            training or evaluation
+
+        Returns
+        -------
+        Dict[str, tf.Tensor]
+            Dict with the metrics values
+        """
+        del x
```

Comment:
Does this actually need to be deleted? Totally possible that it does, but wondering how we know it's safe and necessary to mark it for garbage collection here.

Reply:
Thanks for your review @karlhigley. We have been using standard Keras `Model.compute_metrics()` (after Marc's refactory), which had the same line [here](https://github.com/keras-team/keras/blob/v2.9.0/keras/engine/training.py#L986). I just override that method so that we conditionally update the metrics (each N steps during training or all steps during evaluation) but kept the same signature and inner code.